### DISCUSSION TITLE
do we include non entities in training data for custom spaCy NER model? What does the structure of the .spacy training data file look like?

### DISCUSSION THREAD

**Discussion by fariiaakh** [explosion/spaCy#11391] [Help: Coding & Implementations] [answered]
Hi Everyone,
I am fairly new to the spaCy community and in training customs models in general so I need some help please. 
I am trying to figure out how to train a custom spaCy model for NER and I am using this dataset from kaggle for a test run:
https://www.kaggle.com/code/lenguyendinhkhanh/named-entity-recognition/data

The words which are entities have special tokens corresponding to them and the words which are not entities have a "O" token corresponding to them.

My question is, when I am creating the training data in the spacy format, Should I include both entity and non entity tags in it to give the model an 'better' understanding of what an entity vs a non entity looks like? Or should I only include the labels which actually point to an entity. 

Please note that I am not talking about fine tuning a pretrained NER model. I am talking about creating a fully custom model for a specialized task.  I am just using a toy dataset from Kaggle to understand what to look out for when I start training my actual model.

this is what the toy dataset looks like:

![image](https://user-images.githubusercontent.com/55149721/187210467-3db62fa4-1bba-47f0-8283-0af8dd46539f.png)

**Comment by thomashacker** [comment, marked as answer]
Hello,
you can use the [spacy convert](https://spacy.io/api/cli#convert) command to convert `IOB` files to the `.spacy` training format.
Here are some [example files](https://github.com/explosion/spaCy/tree/master/extra/example_data/ner_example_data) which you can use as a starting point for your data.

**Reply by fariiaakh** [reply]
Thanks a lot edward! my file is currently a csv. The words are in one column and the corresponding tags are in the other column... are there any modules for converting CSVs to IOB files? or any tips you can give me please 🙏

**Reply by fariiaakh** [reply]
oh wait, i didn't realize IOB files were basically TSV files with specialized data: https://datasaurai.gitbook.io/datasaur/supported-formats

**Reply by fariiaakh** [reply]
Hi Edward, 

The documentation says that there is a iob converter  (https://spacy.io/api/cli#convert)  but when i run the command in command prompt, i am getting error that say _cant find converter for 'iob'__ 

please see screenshot: 
![image](https://user-images.githubusercontent.com/55149721/188720547-eca6a78d-7630-4587-a746-2dc203494a26.png)


can you please help me

**Reply by thomashacker** [reply]
Hello, I think the quotes might be the problem. 
Could you try again with `iob` instead of `'iob'`? If that doesn't work, what spaCy version are you using?

**Reply by fariiaakh** [reply]
Thanks a lot Edward! That worked! 🎉🎉🎉  But is there a way to view the converted file? I came across the corpus function here: https://spacy.io/api/data-formats#training and here: https://spacy.io/api/corpus
  but when i tried to read the .spacy file with the corpus function, i keep seeing the file as an object and i can't seem to access its contents. please see below:

`>>> reader = Corpus("nermini.spacy")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Corpus' is not defined
>>> import spacy
>>> reader = Corpus("nermini.spacy")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Corpus' is not defined
>>> reader = spacy.Corpus("nermini.spacy")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'spacy' has no attribute 'Corpus'
>>> from spacy.training import Corpus
>>> reader = spacy.Corpus("nermini.spacy")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'spacy' has no attribute 'Corpus'
>>> reader = Corpus("nermini.spacy")
>>> reader
<spacy.training.corpus.Corpus object at 0x0000016186B4B490>
>>> print(reader)
<spacy.training.corpus.Corpus object at 0x0000016186B4B490>
>>>`

I want to see what the contents of the .spacy file looks like before i start training my custom model
please help 🙏

**Reply by rmitsch** [reply]
Hi @fariiaakh, looks like you haven't imported `Corpus`. Have a look at the [example in the docs](https://spacy.io/api/corpus#init) and add `from spacy.training import Corpus`.

**Reply by rmitsch** [reply]
If you want to look at the contents of your `.spacy` file, I recommend using a `DocBin` however:
```python
import spacy
from spacy.tokens import DocBin

nlp = spacy.blank("en") 
doc_bin = DocBin()
doc_bin.from_disk("nermini.spacy")
for doc in doc_bin.get_docs(nlp.vocab):
    print(doc)
```

**Reply by fariiaakh** [reply]
hey Raphael,

thanks alot for your response! Is there anyways to see the ner tags in the .spacy file, like how i can see them in the json format? like for example:

this is the original iob file: 
![image](https://user-images.githubusercontent.com/55149721/189167706-52cab312-ead4-4abd-84f7-23b891510a96.png)

this is how it looks in the json format from spacy v2:
![image](https://user-images.githubusercontent.com/55149721/189167893-f036a565-f71d-4d99-ac4f-84d8e791e2cf.png)

and when i run the code you provided, i see the words arranged into sentences without any of the ner tags:
![image](https://user-images.githubusercontent.com/55149721/189168789-ef053f3f-3a64-40b8-b40d-2b0a294e9230.png)

is there a way to see the ner tags in the .spacy file, the way i can see them in the json file? 
Basically i want to know what is the format of the data that is contained in the .spacy file?

**Reply by rmitsch** [reply]
You can access the NER tags via the `label_` property - see also the [spaCy docs on Named Entity Recognition](https://spacy.io/usage/linguistic-features#named-entities). For more information on the `.spacy` file I recommend the docs on the [Binary Training Format](https://spacy.io/api/data-formats#binary-training) and [`DocBin`](https://spacy.io/api/docbin).

**Reply by fariiaakh** [reply]
thank you very much Raphael and Edward!

**Reply by fariiaakh** [reply]
i also found the spacy course! i think the tile for this should be higher on the main website😅
https://course.spacy.io/
